---
title: "Create images"
author: "Alex Cope"
date: '2023-06-21'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(phytools)
library(ggtree)
library(cowplot)
library(patchwork)
library(dendextend)
library(ComplexHeatmap)
library(scales)
library(ggpubr)
library(broom)
library(LaplacesDemon,lib.loc="~/R_dev/")

CleanData <- function(phy, data) {
  cleaned <- geiger::treedata(phy,data,warnings=T)
  return(cleaned)
}

createSummaryFileForTargets <- function(mcmc.r.object.file,
                              output = "./summary.tsv")
{
  load(mcmc.r.object.file)
  fit.summary <- as.data.frame(fit$Summary2) %>%
    rownames_to_column("Parameter")
  rm(fit)
  write_tsv(fit.summary,output)
}


mrna.color <- "#EE6677"
prot.color <- "#AA3377"
slope.color <- "#CCBB44"

```

# Plot tree

```{r fig.width=8}
tree <- read.tree("Data/vertlife_mcc_dna_only_node_dated.tre")
tree <- force.ultrametric(tree,method="extend")
plot(tree)
```

```{r}
d <- ggimage::phylopic_uid(tree$tip.label)
d <- d %>%
  mutate(name = case_when(
      name == "Felis_catus" ~ "Cat",
      name == "Bos_taurus" ~ "Cow",
      name == "Canis_lupus" ~ "Dog",
      name == "Equus_caballus" ~ "Horse",
      name == "Homo_sapiens" ~ "Human",
      name == "Macaca_mulatta" ~ "Macaque",
      name == "Ovis_aries" ~ "Sheep",
      name == "Sus_scrofa" ~ "Pig",
      name == "Rattus_norvegicus" ~ "Rat",
      name == "Monodelphis_domestica" ~ "Opossum",
      name == "Oryctolagus_cuniculus" ~ "Rabbit"
    ))
rownames(d) <- d$name

d <- d %>% mutate(uid=case_when(
  name == "Human"~"b8c16fc6-d16b-4fac-8a04-67182448157e",
  name == "Rat"~"53f8f372-adc0-4247-bc66-0075e241fc95",
  T ~ uid) 
)

```

```{r}
mapLatinToCommon <- function(x)
{
  common.names <- c("Opossum","Rabbit","Rat","Human","Macaque","Pig","Cow","Sheep","Horse","Cat","Dog")
  names(common.names) <- c("Monodelphis_domestica",
                           "Oryctolagus_cuniculus",
                           "Rattus_norvegicus",
                           "Homo_sapiens",
                           "Macaca_mulatta",
                           "Sus_scrofa",
                           "Bos_taurus",
                           "Ovis_aries",
                           "Equus_caballus",
                           "Felis_catus",
                           "Canis_lupus"
                           )
  return(common.names[x])
}

tree$tip.label <- mapLatinToCommon(tree$tip.label)
tree.no.human <- drop.tip(tree,"Human")
```

```{r}
node.date <- read.tree("Data/vertlife_mcc_dna_only_node_dated.tre")
tip.date <- read.tree("Data/vertlife_mcc_dna_only_fossil_bd.tre")
ba.tree <- read.nexus("Data/tree_11sp_noGpig.nex")
plot(ba.tree)
plot(tip.date)
plot(node.date)
```




```{r}

data <- read_tsv("Data/mean_se_rna_protein.tsv")
data <- data %>%
  mutate(Species = case_when(
      Species == "Felis_catus" ~ "Cat",
      Species == "Bos_taurus" ~ "Cow",
      Species == "Canis_lupus" ~ "Dog",
      Species == "Equus_caballus" ~ "Horse",
      Species == "Homo_sapiens" ~ "Human",
      Species == "Macaca_mulatta" ~ "Macaque",
      Species == "Ovis_aries" ~ "Sheep",
      Species == "Sus_scrofa" ~ "Pig",
      Species == "Rattus_norvegicus" ~ "Rat",
      Species == "Monodelphis_domestica" ~ "Opossum",
      Species == "Oryctolagus_cuniculus" ~ "Rabbit"
    ))

data.prot.wide <- data %>% 
  dplyr::select(Gene_ID,Species,Mean_Protein) %>%
  pivot_wider(id_cols=Gene_ID,names_from="Species",values_from="Mean_Protein")
data.mrna.wide <- data %>% 
  dplyr::select(Gene_ID,Species,Mean_RNA) %>%
  pivot_wider(id_cols=Gene_ID,names_from="Species",values_from="Mean_RNA")


tree.to.use <- tree.no.human

pairwise.prot <- cor(data.prot.wide[,2:ncol(data.prot.wide)],method="spearman")
pairwise.mrna <- cor(data.mrna.wide[,2:ncol(data.prot.wide)],method="spearman")

pairwise.prot <- pairwise.prot[tree.to.use$tip.label,tree.to.use$tip.label]
pairwise.mrna <- pairwise.mrna[tree.to.use$tip.label,tree.to.use$tip.label]
pairwise.prot[upper.tri(pairwise.prot)] <- pairwise.mrna[upper.tri(pairwise.mrna)]

tree.dend <- as.dendrogram(tree.to.use)

diag(pairwise.prot) <- NA
prot.cor <- Heatmap(pairwise.prot,
        col = rev(hcl.colors(25, palette = "Blues 3")),
        na_col = "black",
        cluster_rows = tree.dend,
        cluster_columns = F,
        name = "Spearman\nCorrelation",
        show_row_names = T,
        row_dend_width = unit(50, "mm"),
        row_names_side = "left",
        row_title="Protein",
        column_title="mRNA",
        )

prot.cor
pdf("Figures/across_species_correlations.pdf",height=7,width=8)
prot.cor
dev.off()
```


```{r}
phyl.dist <- cophenetic.phylo(tree.no.human)
phyl.dist[upper.tri(phyl.dist,diag = T)] <- NA
phyl.dist.long <- as.data.frame(phyl.dist) %>%
  rownames_to_column("Species") %>%
  pivot_longer(-Species,names_to="Other.species",values_to="Distance") %>%
  filter(!is.na(Distance))
pairwise.prot[upper.tri(pairwise.prot,diag = T)] <- NA
pairwise.prot.long <- as.data.frame(pairwise.prot) %>%
  rownames_to_column("Species") %>%
  pivot_longer(-Species,names_to="Other.species",values_to="Correlation") %>%
  filter(!is.na(Correlation)) %>%
  mutate(Measure = "Protein")

pairwise.mrna[upper.tri(pairwise.mrna,diag = T)] <- NA
pairwise.mrna.long <- as.data.frame(pairwise.mrna) %>%
  rownames_to_column("Species") %>%
  pivot_longer(-Species,names_to="Other.species",values_to="Correlation") %>%
  filter(!is.na(Correlation)) %>%
  mutate(Measure = "mRNA")

pairwise <- rbind(pairwise.prot.long,pairwise.mrna.long)


dist.cor <- phyl.dist.long %>%
  left_join(pairwise)
dist.cor.prot <- dist.cor %>% filter(Measure == "Protein")
dist.cor.mrna <- dist.cor %>% filter(Measure == "mRNA")
```

```{r}


div.p <- ggplot(dist.cor,aes(x=Distance,y=Correlation,color=Measure)) +
  geom_point() +
  theme_cowplot() +
  scale_color_manual(values = c("mRNA"=mrna.color,"Protein"=prot.color)) +
  labs(color="Expression type") +
  xlab("Divergence Time (MYA)") +
  ylab("Pairwise Similarity\n(Spearman Correlation)")
div.p

```



```{r fig.width =12,fig.height=6}
p <- ((plot_spacer() | grid.grabExpr(draw(prot.cor),wrap = T,wrap.grobs = T)) | (div.p))  +
  plot_annotation(tag_levels = "A") + 
  plot_layout(widths = c(0,2,1),guides = "collect") 
p

ggsave2("Figures/Figure_2.pdf",p,width=14,height = 5)

```

```{r}

calculatePIC <- function(data,tree)
{
  
  pic.value <- pic(data,tree,var.contrasts = T)
  corr <- cor.test(pic.value[,1],sqrt(pic.value[,2]),method="spearman",use="complete.obs")
  return(pic.value)
}

across.species.corr. <- data %>% 
  group_by(Gene_ID) %>% 
  nest(-Gene_ID) %>% 
  mutate(corr = purrr::map(data,function(x)
    {
      tmp <- x %>% column_to_rownames("Species") %>% 
        dplyr::select(Mean_RNA,Mean_Protein) %>% 
        dplyr::filter(!is.na(Mean_RNA) & !is.na(Mean_Protein)) %>%
        as.matrix()
      cleaned <- CleanData(tree.to.use,tmp)
      x <- cleaned$data
      tmp.tree <- cleaned$phy
      rna.values <- x[,"Mean_RNA"]
      protein.values <- x[,"Mean_Protein"]
      rna.pic <- calculatePIC(rna.values[tmp.tree$tip.label],tree = tmp.tree)
      protein.pic <- calculatePIC(protein.values[tmp.tree$tip.label],tree=tmp.tree)
      cor.test(rna.pic[,1],protein.pic[,1],method="spearman") 
    } %>% tidy())) %>% 
  dplyr::select(-data) %>% 
  unnest(corr,names_sep="_") %>%
  arrange(corr_estimate)


mrna.vs.prot.cor <- ggplot(across.species.corr.,aes(x=corr_estimate)) +
  geom_histogram(color="black",fill="white") +
  ggtitle("Variation in correlation between mRNA\nand protein abundances across species") +
  xlab("Spearman Correlation (PIC)") +
  ylab("Count") +
  theme_cowplot() +
  theme(aspect.ratio=1)
mrna.vs.prot.cor

ggsave2("Figures/mrna_protein_corr_across_species_pic.pdf",mrna.vs.prot.cor)
```


## Comparing models using Deviance Information Criterion

The code immediately below this is what was used locally. This relies on the R objects saved from `LaplacesDemon`. These are too large to upload to the GitHub repo. We have also included code that recalculates DIC from the `summary.tsv` files stored from each run in the code chunk below this one. They give the same answer, but recognize their could be small discrepancies due to rounding, truncation, etc. when the R object summary object is saved as a FLAT file (this is pure speculation on my part...). 
```{r,eval=F}
load("Results/Drop_human/Pop_average_model_all/Protein_selection_strong/2024-10-27_prot_determines_mrna_prior_based_on_vertlife_mcc_node_dated_tree_random_start/rwm_mcmc_1.Rda")
prot.dom.fit <- fit
load("Results/Drop_human/Pop_average_model_all/Protein_selection_weak/2024-10-27_mrna_determines_prot_off_diagonal_sigma_prior_based_on_vertlife_mcc_node_dated_tree_random_start/rwm_mcmc_1.Rda")
mrna.dom.fit <- fit
load("Results/Drop_human/Pop_average_model_all/Independent/2024-10-27_independent_prior_based_on_vertlife_mcc_node_dated_tree_random_start/rwm_mcmc_1.Rda")
indep.fit <- fit

prot.dom.fit.dic <- prot.dom.fit$DIC2[3]
mrna.dom.fit.dic <- mrna.dom.fit$DIC2[3]
indep.fit.dic <- indep.fit$DIC2[3]


prot.dom.fit.dic - mrna.dom.fit.dic
prot.dom.fit.dic - indep.fit.dic
```


```{r}

prot.dom.fit <- read_tsv("Results/Drop_human/Pop_average_model_all/Protein_selection_strong/2024-10-27_prot_determines_mrna_prior_based_on_vertlife_mcc_node_dated_tree_random_start/summary.tsv")
mrna.dom.fit <- read_tsv("Results/Drop_human/Pop_average_model_all/Protein_selection_weak/2024-10-27_mrna_determines_prot_off_diagonal_sigma_prior_based_on_vertlife_mcc_node_dated_tree_random_start/summary.tsv")
indep.fit <- read_tsv("Results/Drop_human/Pop_average_model_all/Independent/2024-10-27_independent_prior_based_on_vertlife_mcc_node_dated_tree_random_start/summary.tsv")

prot.dom.fit <- prot.dom.fit %>% filter(Parameter == "Deviance")
mrna.dom.fit <- mrna.dom.fit %>% filter(Parameter == "Deviance")
indep.fit <- indep.fit %>% filter(Parameter == "Deviance")
mrna.dom.fit <- mrna.dom.fit %>% filter(Parameter == "Deviance")
indep.fit <- indep.fit %>% filter(Parameter == "Deviance")

calculateDICFromSummary <- function(deviance.vec)
{
  return(deviance.vec$Mean + (deviance.vec$SD)^2/2) # DIC is calculated using the suggested formula from Gelman.
}

prot.dom.fit.dic <- calculateDICFromSummary(prot.dom.fit)
mrna.dom.fit.dic <- calculateDICFromSummary(mrna.dom.fit)
indep.fit.dic <- calculateDICFromSummary(indep.fit)


prot.dom.fit.dic - mrna.dom.fit.dic
prot.dom.fit.dic - indep.fit.dic
```

```{r}
dic.scores <- c(prot.dom.fit.dic,mrna.dom.fit.dic,indep.fit.dic)
min.dic <- dic.scores[which.min(dic.scores)]
delta.dic <- min.dic - dic.scores 

dic.df <- data.frame(Model =c("Protein-driven","mRNA-driven","Independent"),
                     DIC = dic.scores,
                     Delta.DIC = delta.dic)# %>%
  #filter(Delta.DIC != 0 )
dic.bar <- ggplot(dic.df,aes(x=Model,y=DIC,fill=Model)) +
  geom_col(color="black") +
  theme_cowplot() +
  ylab("DIC") +
  scale_y_reverse() +
  xlab("") +
  scale_fill_manual(values=c("Protein-driven"=prot.color,"mRNA-driven"=mrna.color,"Independent"="#4477AA")) +
  theme(axis.text.x = element_blank()) +
  geom_bracket(xmin = "Independent", xmax = "Protein-driven", y.position = -60000,label = as.character(round(dic.df$Delta.DIC[3])), tip.length = c(-0.08, -0.12),vjust = 8,inherit.aes=F,label.size=3) +
  geom_bracket(xmin = "mRNA-driven", xmax = "Protein-driven", y.position = -54500,label = as.character(round(dic.df$Delta.DIC[2])), tip.length = c(-0.04, -0.07),vjust=4.5,inherit.aes=F,label.size=3)
dic.bar
```

## Plot histograms of parameter posterior distributions -- will require the R objects 

This code cannot be run using the `summary.tsv` files, but requires the R object files output by `LaplacesDemon`. Please reach out to one of the authors so we can get you these files. You can also recreate the runs yourself using the code available in `R_scripts` and the `README.md` files for each run. This should generate results similar to ours (and please let us know if they do not). 

```{r}

alpha.b <- prot.dom.fit$Posterior2[,1:5]

hlf.trace <- apply(alpha.b,1,function(x)
  {
   tmp <- matrix(c(exp(x[1]),0,-exp(x[2])*x[3],exp(x[2])),nrow=2)
   lambda<-eigen(tmp)$values
   phyhalflife<-log(2)/Re(lambda)
   phyhalflife
})
hlf.trace <- t(hlf.trace) 
colnames(hlf.trace) <- c("mRNA","Protein")
hlf.df <- hlf.trace %>% 
  as.data.frame() %>% 
  pivot_longer(cols=everything(),names_to="Abundance",values_to="Halflife")

mean.hlf <- apply(hlf.trace,MARGIN = 2,FUN = mean)
ci.hlf <- p.interval(hlf.trace)

```

```{r}

param.names <- colnames(alpha.b)
alpha.b.nat.scale <- sapply(param.names,function(x)
  {
    if (x != "Q")
    {
      exp(alpha.b[,x])
    } else {
      alpha.b[,x]
    }
})

mean.nat.scale <- apply(alpha.b.nat.scale,MARGIN = 2,FUN = mean)
ci.nat.scale <- p.interval(alpha.b.nat.scale)
ci.nat.scale

```

```{r fig.width=8,fig.height=3}

alpha.param <- exp(alpha.b[,c(1,2)])
colnames(alpha.param) <- c("mRNA","Protein")
alpha.df <- alpha.param %>% as.data.frame() %>% pivot_longer(cols=everything(),
                                                         names_to="Abundance",
                                                         values_to="Selection")
alpha.plot <- ggplot(alpha.df,aes(x=Selection,fill=Abundance)) +
  geom_histogram(color="black",position="identity",alpha=0.8) +
  xlab("Rate of adaptation") +
  ylab("Count") +
  labs(fill="Expression type") +
  scale_fill_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color))+
  theme_cowplot()

alpha.plot


hlf.plot <- ggplot(hlf.df,aes(x=Halflife,fill=Abundance)) +
  geom_histogram(color="black",alpha=0.8,position="identity") +
  xlab("Phylogenetic half-life (MY)") +
  labs(fill="Expression type") +
  ylab("Count") +
  scale_fill_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  theme_cowplot()

hlf.plot


slope.df <- data.frame(Distribution="Posterior",Q=alpha.b[,"Q"])

slope.plot <- ggplot(slope.df,aes(x=Q)) +
  geom_histogram(fill=slope.color,color="black",position="identity") +
  xlab(expression(atop("Evolutionary slope a"[R],"mRNA per protein"))) +
  ylab("Count") +
  cowplot::theme_cowplot()
slope.plot

tau.trace <- data.frame(mRNA=exp(alpha.b[,"Sigma_Y"]),Protein=exp(alpha.b[,"Sigma_X"]))
tau.df <- tau.trace %>% 
  pivot_longer(everything(),names_to="Abundance",values_to="Sigma")

tau.plot <- ggplot(tau.df,aes(x=Sigma,fill=Abundance)) +
  geom_histogram(color="black",position="identity",alpha=0.8) +
  xlab("Rate of drift") +
  ylab("Count") +
  labs(fill="Expression type") +
  scale_fill_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color))+
  theme_cowplot()
tau.plot


```

```{r fig.width=10, fig.height=10}
total <- (dic.bar + alpha.plot + plot_layout(widths = c(1,1.9))) / (slope.plot | tau.plot) + plot_annotation(tag_levels ="A")
ggsave2("Figures/parameter_histograms_and_dic_plot.pdf",total,width=10,height = 7.25)
```




### Check for consistency across runs - resume code that can be run without the R objects

Here, I will be looking at the runs where the human species is dropped from our analysis.


```{r}

createSummaryFile <- function(file.path.to.runs)
{
  r.object.files <- list.files(file.path.to.runs,recursive = T,pattern="rwm_mcmc_1.Rda",full.names = T)

  output <- str_split(r.object.files ,"/")
  output <- lapply(output,function(x){
    filepath <- paste(x[1:(length(x) - 1)],collapse="/")
    filepath <- paste(filepath,"summary.tsv",sep="/")
    filepath
  })
  
  
  purrr::map2(r.object.files ,output,createSummaryFileForTargets)
}

createSummaryFile("Results/Drop_human/Pop_average_model_all/Protein_selection_strong")    
createSummaryFile("Results/Drop_human/Pop_average_model_all/Protein_selection_weak") 
createSummaryFile("Results/Drop_human/Pop_average_model_all/Independent")



```


```{r}

# model should be "Protein Driven", "mRNA Driven", "Independent"
convertParameterSymbolsToPaperSymbols <- function(parameter.names,model = "Protein Driven")
{
  prot.dom.list <- list("H_Y"="alpha[R]","H_X"="alpha[P]","Q" = "a[R]","Sigma_Y"="tau[R]","Sigma_X"="tau[P]")
  mrna.dom.list <-list("H_Y"="alpha[P]","H_X"="alpha[R]","C"="c","Q" = "a[P]","Sigma_Y"="tau[P]","Sigma_X"="tau[R]")
  indep.list <- list("H_Y"="alpha[R]","H_X"="alpha[P]","Sigma_Y"="tau[R]","Sigma_X"="tau[P]")
  if (model == "Protein Driven")
  {
    name.map <- prot.dom.list[parameter.names]
  } else if (model == "mRNA Driven") {
    name.map <- mrna.dom.list[parameter.names]
  } else if (model == "Independent") {
    name.map <- indep.list[parameter.names]
  } else{
    print("Model name not one of the three options: Protein Driven, mRNA Driven, or Independent. Returning NA")
    name.map <- NA
  }
  return(unname(unlist(name.map)))
}



compareParameterEstimatesAcrossRuns <- function(file.path.to.summary.files,
                                                model="Protein Driven")
{
  prot.sel.strong.summary.files <- list.files(file.path.to.summary.files,
                                            recursive = T,
                                            pattern = "summary.tsv",
                                            full.names = T) 

  names(prot.sel.strong.summary.files) <- paste("Run",1:length(prot.sel.strong.summary.files),sep="_")
  
  prot.sel.strong.param <- lapply(prot.sel.strong.summary.files,read_tsv) %>%
    bind_rows(.id="Run")
  
  
  param <- c("H_X","H_Y","Q","C","Sigma_X","Sigma_Y")
  prot.sel.strong.pop.param <- prot.sel.strong.param %>% 
    filter(Parameter %in% param) %>%
    mutate(Parameter = convertParameterSymbolsToPaperSymbols(Parameter,model=model))

  prot.sel.strong.pop.param <- prot.sel.strong.pop.param %>%
    mutate(Run = case_when(
      Run == "Run_1" ~ "Ba et al. Tree 1",
      Run == "Run_2" ~ "Ba et al. Tree 2",
      Run == "Run_3" ~ "Ba et al. Tree 3",
      Run == "Run_4" ~ "VertLife MCC Tip Dated Tree 1",
      Run == "Run_5" ~ "VertLife MCC Node Dated Tree 1",
      Run == "Run_6" ~ "VertLife MCC Node Dated Tree 2",
      Run == "Run_7" ~ "VertLife MCC Node Dated Tree 3",
      Run == "Run_8" ~ "VertLife MCC Node Dated Tree 4",
    ))
  
  p <- ggplot(prot.sel.strong.pop.param,aes(x=Parameter,y=Mean,color=Run)) +
    geom_point(position=position_dodge(width=0.5),size=2) +
    geom_errorbar(aes(ymin=LB,ymax=UB),position=position_dodge(width=0.5),width=0.2) +
    theme_cowplot() +
    ylab("Posterior Mean") +
    ggtitle(model) +
    scale_x_discrete(labels=label_parse()) +
    ylim(c(-5,1))
  p
}

prot.sel.strong.p <- compareParameterEstimatesAcrossRuns("Results/Drop_human/Pop_average_model_all/Protein_selection_strong/",
                                    model="Protein Driven")

prot.sel.weak.p <- compareParameterEstimatesAcrossRuns("Results/Drop_human/Pop_average_model_all/Protein_selection_weak/",
                                    model="mRNA Driven")

indep.p <- compareParameterEstimatesAcrossRuns("Results/Drop_human/Pop_average_model_all/Independent//",
                                    model = "Independent")


```

```{r fig.width=10,fig.height=3}
comb.plot <- (prot.sel.strong.p | prot.sel.weak.p | indep.p) + 
  plot_annotation(tag_levels = "A") +
  plot_layout(guides="collect")
comb.plot

ggsave2("Figures/convergence_across_mcmc_runs.pdf",comb.plot,width=15,height=5)

```

### Assess impact of gene expression

The impact of gene expression on the evolutionary patterns observed across species of orthologous proteins is well-established. 

```{r}


getDIC <- function(filepath.mcmc.object)
{
  load(filepath.mcmc.object)
  return(data.frame(Deviance = fit$DIC1[1],Eff.Num.Param=fit$DIC1[2],DIC = fit$DIC1[3]))
}

getSummaryTable <- function(filepath.mcmc.object,adapt=F)
{
  load(filepath.mcmc.object)
  if (adapt==T)
  {
    fit <- adapt.fit
  }
  fit.summary <- as.data.frame(fit$Summary2) %>%
    rownames_to_column("Parameter") %>%
    as.data.frame()
  return(fit.summary)
}

extractRunNameFromFilePath <- function(filepath,pattern="[0-9]{1,2}-[0-9]{2,3}/[A-Za-z_]+")
{
 return(str_extract(filepath,pattern=pattern))
}

createDICTable <- function(filepaths.mcmc.obj,pattern="[0-9]{1,2}-[0-9]{2,3}/[A-Za-z_]+")
{
  names(filepaths.mcmc.obj) <- extractRunNameFromFilePath(mcmc.rda,pattern=pattern)
  dic.table <- lapply(filepaths.mcmc.obj,getDIC) %>%
    bind_rows(.id="Data-Model_Fit")
  dic.table <- dic.table %>%
    separate(`Data-Model_Fit`,sep="/",into = c("Data","Model_Fit"))
  return(dic.table)
}

createParamTableFromRobjects <- function(filepaths.mcmc.obj,pattern="[0-9]{1,2}-[0-9]{2,3}/[A-Za-z_]+",adapt=F)
{
  names(filepaths.mcmc.obj) <- extractRunNameFromFilePath(mcmc.rda,pattern=pattern)
  param.table <- lapply(filepaths.mcmc.obj,getSummaryTable,adapt=adapt) %>%
    bind_rows(.id="Data-Model_Fit")
  param.table <- param.table %>%
    separate(`Data-Model_Fit`,sep="/",into = c("Data","Model_Fit"))
  return(param.table)
}

createParamTableFromSummary <- function(file.path.summary,pattern="[0-9]{1,2}-[0-9]{2,3}/[A-Za-z_]+")
{
  names(file.path.summary) <- extractRunNameFromFilePath(file.path.summary,pattern=pattern)
  param.table <- lapply(file.path.summary,read_tsv,col_types=cols()) %>%
    bind_rows(.id="Data-Model_Fit")
  param.table <- param.table %>%
    separate(`Data-Model_Fit`,sep="/",into = c("Data","Model_Fit"))
  return(param.table)
}




toplevel <- list.files(path="Results/Drop_human/Bin_based_on_paxdb_human_prot_abund/",pattern = "^[0-9]{1,2}-[0-9]{2,3}$",include.dirs = T)
prot.strong.summary <- list.files(path=file.path("Results/Drop_human/Bin_based_on_paxdb_human_prot_abund/",toplevel,"Protein_selection_strong","2024-11-01_prot_determines_mrna_prior_based_on_vertlife_mcc_node_dated_tree_random_start"),pattern="summary.tsv",recursive = T,full.names = T)

prot.dom.param.table <- createParamTableFromSummary(prot.strong.summary) %>%
  mutate(Bin.order = case_when(
             Data == "0-10" ~ 1,
             Data == "10-20" ~ 2,
             Data == "20-30" ~ 3,
             Data == "30-40" ~ 4,
             Data == "40-50" ~ 5,
             Data == "50-60" ~ 6,
             Data == "60-70" ~ 7,
             Data == "70-80" ~ 8,
             Data == "80-90" ~ 9,
             Data == "90-100" ~ 10
           ))

```





```{r}
param.table.selection <- prot.dom.param.table %>%
    filter((Parameter == "H_X" | Parameter == "H_Y")  & Model_Fit == "Protein_selection_strong") %>% 
    mutate(Parameter = ifelse(Parameter == "H_X","Protein","mRNA"))



selec.p <- ggplot(param.table.selection,aes(x=Data,y=Mean,color=Parameter)) +
  geom_point(position=position_dodge(width=0.5)) +
  geom_errorbar(aes(ymin=LB,ymax=UB),width=0,position=position_dodge(width=0.5)) +
  theme_cowplot() +
  stat_cor(method="spearman",aes(x=Bin.order,y=Mean,label=..r.label..),show.legend = F) +
  xlab("Gene Expression Bin") +
  ylab("Rate of adaptation (Log scale)") +
  scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


param.table.tau <- prot.dom.param.table %>%
  filter((Parameter == "Sigma_X" | Parameter == "Sigma_Y")  & Model_Fit == "Protein_selection_strong") %>%
  mutate(Parameter = ifelse(Parameter == "Sigma_X","Protein","mRNA"))


tau.p <- ggplot(param.table.tau,aes(x=Data,y=Mean,color=Parameter)) +
  geom_point(position=position_dodge(width=0.5)) +
  geom_errorbar(aes(ymin=LB,ymax=UB),width=0,position=position_dodge(width=0.5)) +
  stat_cor(method="spearman",aes(x=Bin.order,y=Mean,label=..r.label..),show.legend = F) +
  theme_cowplot() +
  xlab("Gene Expression Bin") +
  ylab("Rate of drift (Log scale)") +
  scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


param.table.q <- prot.dom.param.table %>%
  filter(Parameter == "Q" & Model_Fit == "Protein_selection_strong")
q.p <- ggplot(param.table.q,aes(x=Data,y=Mean)) +
  geom_point(color=slope.color) +
  geom_errorbar(aes(ymin=LB,ymax=UB),width=0,color=slope.color) +
  stat_cor(method="spearman",aes(x=Bin.order,y=Mean,label=..r.label..),show.legend = F) +
  theme_cowplot() +
  xlab("Gene Expression Bin") +
  ylab(expression("Evolutionary slope a"[R])) +
  scale_x_discrete(labels=c("Lowest",rep("",8),"Highest"))
  #theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))



```


```{r fig.width=8,fig.height=4}
gene.expr.effect <- (selec.p | tau.p | q.p) + 
  plot_annotation(tag_level = "A") + 
  plot_layout(guides = "collect")
gene.expr.effect

ggsave2("Figures/impact_of_expression_on_protein_dominant_model_parameters_humans.pdf",gene.expr.effect,width=15,height = 5)

```



```{r}

hsapien.prot <- read_tsv("Data/PaxDB/paxdb_human_whole_organism_prot_abundance_integrated.tsv",comment="#") %>%
  mutate(string_external_id = str_remove(string_external_id,"9606."))

ensembl.prot.to.gene <- read_tsv("Data/PaxDB/hsapien_gene_mapping.tsv")

hsapien.prot <- hsapien.prot %>%
  left_join(ensembl.prot.to.gene,by=c("string_external_id"="ensembl_peptide_id"))

ba.data <- read_tsv("Data/mean_se_rna_protein.tsv")
ba.genes <- ba.data %>%
  dplyr::select(Gene_ID) %>%
  deframe() %>%
  unique()

hsapien.prot.filt <- hsapien.prot %>%
  filter(ensembl_gene_id %in% ba.genes) %>% 
  group_by(ensembl_gene_id) %>% 
  summarize(abundance = sum(abundance))

missing.ids <- ba.genes[which(!ba.genes %in% hsapien.prot.filt$ensembl_gene_id)]

hsapien.prot.filt <- hsapien.prot.filt %>%
  mutate(log10.abundance = log10(abundance),
         Bin = Hmisc::cut2(log10.abundance,g = 10))


eqtl.human <- read_tsv("Data/GTEx/eQTL_summary.tsv") %>%
  dplyr::select(gene_id,Mean.abs.effect) %>%
  dplyr::rename(Mean.Abs.eQTL.Effect = Mean.abs.effect)
number.enhancers.human <- read_csv("Data/CRE/EnhancerAtlas/human_keratinocyte_enhancer_counts_per_gene.csv") %>% 
  dplyr::rename(Num.Enhancers.Human = Num.Enhancers)
number.cres.human <- read_csv("Data/CRE/Human_cell_atlas/human_cell_atlas_cre.csv") %>% 
  dplyr::rename(Num.tCREs.Human = Num.Enhancers)
number.enhancers.mouse <- read_csv("Data/CRE/EnhancerAtlas/mouse_kidney_enhancer_counts_per_gene_w_human_id.csv") %>% 
  dplyr::rename(Num.Enhancers.Mouse = Num.Enhancers) %>%
  dplyr::select(-Gene_ID)
number.enhancers.fly <- read_csv("Data/CRE/EnhancerAtlas/fly_bg3c2_enhancer_counts_per_gene_w_human_id.csv") %>%
  dplyr::rename(Num.Enhancers.Fly = Num.Enhancers) %>%
  dplyr::select(-Gene_ID)
te.human <- read_tsv("Data/Translation_efficiency/ribo_te/human_te_l29.tsv") %>%
  mutate(TE = Ribo_TPM/RNA_TPM)

phylop <- read_tsv("Data/Translation_efficiency/conservation/2024-11-11_human_utr5_phylop.tsv") %>%
  dplyr::select(ensembl_gene_id,agg_PGS) %>%
  dplyr::rename(Mean.PhyloP = agg_PGS)

phast <- read_tsv("Data/Translation_efficiency/conservation/2024-11-11_human_utr5_phast.tsv") %>%
  dplyr::select(ensembl_gene_id,agg_PGS) %>%
  dplyr::rename(Mean.Phast = agg_PGS)

utr5.length <- read_tsv("Data/Translation_efficiency/utr5_length/human_vs_mouse_utr5_length_level_all_by_all_princ_isoforms.tsv")
  
human.utr5.length <- utr5.length %>%
  dplyr::select(Human,Max.UTR5.Length_Human) %>%
  filter(!is.na(Max.UTR5.Length_Human)) %>%
  filter(Human %in% ba.data$Gene_ID) %>%
  distinct(Human,.keep_all = T)

mouse.utr5.length <- utr5.length %>%
  dplyr::select(Gene,Human,Max.UTR5.Length_Mouse) %>%
  filter(!is.na(Max.UTR5.Length_Mouse)) %>% 
  filter(Human %in% ba.data$Gene_ID) %>%
  dplyr::select(-Gene) %>%
  distinct(Human,.keep_all = T)


hsapien.prot.filt <- hsapien.prot.filt %>%
  left_join(eqtl.human,by=c("ensembl_gene_id"="gene_id")) %>%
  left_join(number.enhancers.human,by=c("ensembl_gene_id"="Gene_ID")) %>%
  left_join(number.cres.human,by=c("ensembl_gene_id"="Gene_ID")) %>%
  left_join(number.enhancers.mouse,by=c("ensembl_gene_id"="Human")) %>%
  left_join(number.enhancers.fly,by=c("ensembl_gene_id"="Human")) %>%
  left_join(te.human,by=c("ensembl_gene_id"="Gene_ID")) %>%
  left_join(human.utr5.length,by=c("ensembl_gene_id"="Human")) %>%
  left_join(mouse.utr5.length,by=c("ensembl_gene_id"="Human")) %>%
  left_join(phylop,by=c("ensembl_gene_id")) %>%
  left_join(phast,by="ensembl_gene_id") %>%
  mutate(Log.TE = log(TE)) %>%
  filter(!is.na(Bin)) 

human.bin.features <- hsapien.prot.filt %>%
  group_by(Bin) %>%
  summarize(Median.Abs.eQTL.Effect=median(Mean.Abs.eQTL.Effect,na.rm=T),
            Median.Num.Enhancers.Human=median(Num.Enhancers.Human,na.rm=T),
            Median.Num.tCREs.Human=median(Num.tCREs.Human,na.rm=T),
            Median.Num.Enhancers.Mouse=median(Num.Enhancers.Mouse,na.rm=T),
            Median.Num.Enhancers.Fly=median(Num.Enhancers.Fly,na.rm=T),
            Median.UTR5.Human = median(Max.UTR5.Length_Human,na.rm=T),
            Median.UTR5.Mouse = median(Max.UTR5.Length_Mouse,na.rm=T),
            Median.PhyloP = median(Mean.PhyloP,na.rm=T),
            Median.Phast = median(Mean.Phast,na.rm=T),
            Median.Log.TE = median(Log.TE,na.rm=T)
            ) %>%
  mutate(Bin = c("0-10","10-20","20-30","30-40","40-50","50-60","60-70","70-80","80-90","90-100"))




param.table.w.enh.eqtl <- prot.dom.param.table %>%
    filter((Parameter %in% c("H_Y","H_X","Q","Sigma_Y","Sigma_X"))  & Model_Fit == "Protein_selection_strong") %>%
    mutate(Abundance = case_when(
      Parameter %in% c("H_X","Sigma_X") ~ "Protein",
      Parameter %in% c("H_Y","Sigma_Y") ~ "mRNA",
      TRUE ~ "Slope"
    ),
    Parameter.type = case_when(
      Parameter %in% c("H_X","H_Y") ~ "Selection",
      Parameter %in% c("Sigma_X","Sigma_Y") ~ "Drift",
      TRUE ~ "Slope"
    ))

param.table.w.enh.eqtl <- param.table.w.enh.eqtl %>%
  left_join(human.bin.features,by=c("Data" = "Bin"))
```

```{r,fig.height=7,fig.width=7}

hsapien.prot.filt.long <- hsapien.prot.filt %>%
  dplyr::select(-log10.abundance,-Tissue,-Ribo_TPM,-RNA_TPM,-Log.TE) %>%
  pivot_longer(-c(ensembl_gene_id,Bin,abundance),names_to="Feature",values_to="Value")

feature.vs.ge.scatter <- ggplot(hsapien.prot.filt.long,aes(x=abundance,y=Value)) +
  geom_point(alpha=0.7) +
  xlab("Human Protein Abundance") +
  stat_cor(method="spearman",label.sep="\n") +
  theme_cowplot() +
  scale_x_log10() +
  facet_wrap(~Feature,scales="free")

feature.vs.ge.scatter


feature.vs.ge.violin <- ggplot(hsapien.prot.filt.long,aes(x=Bin,y=Value)) +
  geom_violin(draw_quantiles = c(0.25,0.5,0.75)) +
  xlab("Human Protein Abundance Bin") +
  theme_cowplot() +
  scale_x_discrete(labels=c("Lowest",rep("",8),"Highest"))+
  #scale_y_log10() +
  facet_wrap(~Feature,scales="free")

feature.vs.ge.violin

ggsave2("Figures/paxdb_human_protein_vs_feature_scatter.pdf",plot=feature.vs.ge.scatter,width = 11,height=11)

ggsave2("Figures/paxdb_human_protein_bins_vs_feature_violin.pdf",plot=feature.vs.ge.violin,width = 11,height=11)

```



## Essential vs. non-essential

```{r}
toplevel <- list.files(path="Results/Drop_human/Essential_vs_non_essential/",pattern = "[A-Za-z_]+",include.dirs = T)
mcmc.rda <- list.files(path=file.path("Results/Drop_human/Essential_vs_non_essential",toplevel,"Protein_selection_strong","2024-11-01_prot_determines_mrna_prior_based_on_vertlife_mcc_node_dated_tree_random_start"),pattern="rwm_mcmc_1.Rda",recursive = T,full.names = T)

ess.param.table <- createParamTable(mcmc.rda,pattern="(Essential|Non_essential)/[A-Za-z_]+")

```


```{r fig.width=5,fig.height=5}
param.table.ess<- ess.param.table %>%
    filter((Data == "Essential" & Model_Fit == "Protein_selection_strong")|(Data == "Non_essential" & Model_Fit == "Protein_selection_strong")) %>%
    filter((Parameter == "H_X" | Parameter == "H_Y")) %>%
    mutate(Parameter = ifelse(Parameter == "H_X","Protein","mRNA"),
           Data = ifelse(Data == "Non_essential","Non-essential","Essential")) %>%
   dplyr::rename(Abundance = Parameter)

ess.p <- ggplot(param.table.ess,aes(x=Data,y=Mean,color=Abundance)) +
    geom_point(position=position_dodge(width=0.5)) +
    geom_errorbar(aes(ymin=LB,ymax=UB),width=0,position=position_dodge(width=0.5)) +
    theme_cowplot() +
    xlab("") +
    labs(color="Expression type") +
    ylab("Rate of adaptation (Log scale)") +
    scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color))
ess.p

```

```{r}
load("Results/Drop_human/Essential_vs_non_essential/Essential/Protein_selection_strong/2024-11-01_prot_determines_mrna_prior_based_on_vertlife_mcc_node_dated_tree_random_start/rwm_mcmc_1.Rda")
ess.fit <- fit

load("Results/Drop_human/Essential_vs_non_essential/Non_essential/Protein_selection_strong/2024-11-01_prot_determines_mrna_prior_based_on_vertlife_mcc_node_dated_tree_random_start/rwm_mcmc_1.Rda")
non.ess.fit <- fit

ess.trace <- exp(ess.fit$Posterior2[,c("H_Y","H_X")])
non.ess.trace <- exp(non.ess.fit$Posterior2[,c("H_Y","H_X")])
  
mean.ess <- apply(ess.trace,MARGIN = 2,FUN = mean)
mean.non.ess <- apply(non.ess.trace,MARGIN = 2,FUN = mean)
ci.ess <- p.interval(ess.trace)
ci.non.ess  <- p.interval(non.ess.trace)

mean.ess
ci.ess
mean.non.ess
ci.non.ess
```


## Haploinsufficient vs. Haplosufficient

```{r}
toplevel <- list.files(path="Results/Drop_human/Haplosufficient_vs_insufficient/",pattern = "[A-Za-z_]+",include.dirs = T)
mcmc.rda <- list.files(path=file.path("Results/Drop_human/Haplosufficient_vs_insufficient/",toplevel,"Protein_selection_strong","2024-11-01_prot_determines_mrna_prior_based_on_vertlife_mcc_node_dated_tree_random_start"),pattern="rwm_mcmc_1.Rda",recursive = T,full.names = T)

haplo.param.table <- createParamTable(mcmc.rda,pattern="(Haplosufficient|Haploinsufficient)/[A-Za-z_]+")

```


```{r fig.width=6,fig.height=5}
param.table.haplo<- haplo.param.table %>%
    filter(Model_Fit == "Protein_selection_strong") %>%
    filter((Parameter == "H_X" | Parameter == "H_Y")) %>%
    mutate(Parameter = ifelse(Parameter == "H_X","Protein","mRNA")) %>%
   dplyr::rename(Abundance = Parameter)

haplo.p <- ggplot(param.table.haplo,aes(x=Data,y=Mean,color=Abundance)) +
    geom_point(position=position_dodge(width=0.5)) +
    geom_errorbar(aes(ymin=LB,ymax=UB),width=0,position=position_dodge(width=0.5)) +
    theme_cowplot() +
    xlab("") +
    labs(color="Expression type") +
    ylab("Rate of adaptation (Log scale)") +
    scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color))
haplo.p

ggsave2("Figures/haploinsufficient.pdf",haplo.p,width = 6,height = 5)

```


## Gene expression and gene architecture

```{r}
sel.vs.gene.exp.p <- ggplot(param.table.w.enh.eqtl %>% filter(Parameter.type=="Selection"),aes(x=Data,y=Mean,color=Abundance)) +
    geom_point(position=position_dodge(width=0.5)) +
    geom_errorbar(aes(ymin=LB,ymax=UB),width=0,position=position_dodge(width=0.5)) +
    stat_cor(method="spearman",aes(x=Bin.order,y=Mean,label=..r.label..),show.legend = F) +
    theme_cowplot() +
    scale_x_discrete(labels=c("Lowest",rep("",8),"Highest"))+
    xlab("Gene Expression Bin") +
    ylab("Rate of adaptation (Log scale)") +
    labs(color="Expression type") +
    scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) 
    #theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
sel.vs.gene.exp.p



sel.vs.eqtl <- ggplot(param.table.w.enh.eqtl %>% filter(Parameter.type=="Selection"),aes(x=Median.Abs.eQTL.Effect,y=Mean,color=Abundance)) +
  geom_point() +
  xlab("Median magnitude of eQTL effect") +
  ylab("Rate of adaptation (Log scale)") +
  geom_errorbar(aes(ymin=LB,ymax=UB),width=0) +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  theme_cowplot() +
  labs(color="Expression type") +
  facet_wrap(~Abundance)


sel.vs.eqtl

ggsave2("Figures/rate_of_adapt_vs_eQTL.pdf",plot=sel.vs.eqtl,height = 5)
```


### Cis-regulatory elements and evolutionary dynamics



#### Humans

```{r}
sel.vs.enh <- ggplot(param.table.w.enh.eqtl %>% filter(Parameter.type=="Selection"),aes(x=Median.Num.Enhancers.Human,y=Mean,color=Abundance)) +
  geom_point() +
  xlab("Median number of enhancers per gene") +
  ylab("Rate of adaptation (Log scale)") +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  geom_errorbar(aes(ymin=LB,ymax=UB), width=0) +
  scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  theme_cowplot() + 
  labs(color="Expression type") +
  facet_wrap(~Abundance)

sel.vs.enh

sel.vs.cre <- ggplot(param.table.w.enh.eqtl %>% filter(Parameter.type=="Selection"),aes(x=Median.Num.tCREs.Human,y=Mean,color=Abundance)) +
  geom_point() +
  xlab("Median number of tCREs per gene") +
  ylab("Rate of adaptation (Log scale)") +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  geom_errorbar(aes(ymin=LB,ymax=UB), width=0) +
  scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  theme_cowplot() + 
  labs(color="Expression type") +
  facet_wrap(~Abundance) +
  ggtitle("Human (single-cell 5'-RNA-seq cell tCREs)")

sel.vs.cre

tau.w.enh.eqtl <- param.table.w.enh.eqtl %>%
  filter(Parameter.type == "Drift") %>%
  group_by(Data) %>%
  mutate(Sigma.Diff = Mean - Mean[Abundance == "Protein"]) %>%
  filter(Abundance == "mRNA")

tau.diff.vs.enh.human <- ggplot(tau.w.enh.eqtl,aes(x=Median.Num.Enhancers.Human,y=Sigma.Diff)) +
  geom_point() +
  xlab("Median number of enhancers per gene") +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  ylab(expression("Log("*tau[R]*") - Log("*tau[P]*")")) +
  theme_cowplot()
tau.diff.vs.enh.human

tau.diff.vs.cre.human <- ggplot(tau.w.enh.eqtl,aes(x=Median.Num.tCREs.Human,y=Sigma.Diff)) +
  geom_point() +
  xlab("Median number of tCREs per gene") +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  ylab(expression("Log("*tau[R]*") - Log("*tau[P]*")")) +
  theme_cowplot()
tau.diff.vs.cre.human


```


#### Mouse and Fly

```{r fig.height=10,fig.width=10}

tau.diff.vs.enh.mouse <- ggplot(tau.w.enh.eqtl,aes(x=Median.Num.Enhancers.Mouse,y=Sigma.Diff)) +
  geom_point() +
  xlab("Median number of enhancers per gene") +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  ylab(expression("Log("*tau[R]*") - Log("*tau[P]*")")) +
  theme_cowplot() 

tau.diff.vs.enh.fly <- ggplot(tau.w.enh.eqtl,aes(x=Median.Num.Enhancers.Fly,y=Sigma.Diff)) +
  geom_point() +
  xlab("Median number of enhancers per gene") +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  ylab(expression("Log("*tau[R]*") - Log("*tau[P]*")")) +
  theme_cowplot() 



sel.vs.enh.mouse <- ggplot(param.table.w.enh.eqtl %>% filter(Parameter.type=="Selection"),aes(x=Median.Num.Enhancers.Mouse,y=Mean,color=Abundance)) +
  geom_point() +
  xlab("Median number of enhancers per gene") +
  ylab("Rate of adaptation (Log scale)") +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  geom_errorbar(aes(ymin=LB,ymax=UB),width=0) +
  scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  theme_cowplot() + 
  labs(color="Expression type") +
  facet_wrap(~Abundance) +
  ggtitle("Mouse")

sel.vs.enh.fly <- ggplot(param.table.w.enh.eqtl %>% filter(Parameter.type=="Selection"),aes(x=Median.Num.Enhancers.Fly,y=Mean,color=Abundance)) +
  geom_point() +
  xlab("Median number of enhancers per gene") +
  ylab("Rate of adaptation (Log scale)") +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  geom_errorbar(aes(ymin=LB,ymax=UB),width=0) +
  scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  theme_cowplot() + 
  labs(color="Expression type") +
  facet_wrap(~Abundance) +
  ggtitle("Fly")

mouse.fly.enh <- (sel.vs.enh.mouse + tau.diff.vs.enh.mouse + sel.vs.enh.fly + tau.diff.vs.enh.fly + sel.vs.cre + tau.diff.vs.cre.human) + plot_layout(guides="collect",nrow=3)
mouse.fly.enh
ggsave2("Figures/mouse_fly_enhancer.pdf",mouse.fly.enh,width = 9,height=9)

```

### mRNA secondary structure divergence

#### Based on Appris Princial Isoforms

```{r}
hsapien.prot <- read_tsv("Data/PaxDB/paxdb_human_whole_organism_prot_abundance_integrated.tsv",comment="#") %>%
  mutate(string_external_id = str_remove(string_external_id,"9606."))

ensembl.prot.to.gene <- read_tsv("Data/PaxDB/hsapien_gene_mapping.tsv")

hsapien.prot <- hsapien.prot %>%
  left_join(ensembl.prot.to.gene,by=c("string_external_id"="ensembl_peptide_id"))

ba.data <- read_tsv("Data/mean_se_rna_protein.tsv")
ba.genes <- ba.data %>%
  dplyr::select(Gene_ID) %>%
  deframe() %>%
  unique()

hsapien.prot.filt <- hsapien.prot %>%
  filter(ensembl_gene_id %in% ba.genes) %>% 
  group_by(ensembl_gene_id) %>% 
  summarize(abundance = sum(abundance))

missing.ids <- ba.genes[which(!ba.genes %in% hsapien.prot.filt$ensembl_gene_id)]

hsapien.prot.filt <- hsapien.prot.filt %>%
  mutate(log10.abundance = log10(abundance),
         Bin = Hmisc::cut2(log10.abundance,g = 10))


start.codon.mrna.struct <- read_tsv("Data/Translation_efficiency/mrna_secondary_struct/human_vs_mouse_start_codon_ss_transcript_id_appris_1_mane_select.tsv") %>%
  mutate(SS.Divergence = abs(CDS_SS_EFE_Human - CDS_SS_EFE_Mouse)) %>%
  filter(!is.na(SS.Divergence) & !is.na(ensembl_gene_id))

hsapien.prot.filt.ss.div <- hsapien.prot.filt %>%
  dplyr::right_join(start.codon.mrna.struct,by="ensembl_gene_id") %>%
  filter(!is.na(Bin))

bin.ss.div <- hsapien.prot.filt.ss.div %>%
  group_by(Bin,Distance) %>%
  summarize(Median.SS.Diverge = median(SS.Divergence,na.rm=T))

new.labels <- c("0-10","10-20","20-30","30-40","40-50","50-60","60-70","70-80","80-90","90-100")
names(new.labels) <- levels(bin.ss.div$Bin)
bin.ss.div$Bin <- unname(new.labels[bin.ss.div$Bin])

param.table.w.ss <- prot.dom.param.table %>%
    filter((Parameter %in% c("H_Y","H_X","Q","Sigma_Y","Sigma_X"))  & Model_Fit == "Protein_selection_strong") %>%
    mutate(Abundance = case_when(
      Parameter %in% c("H_X","Sigma_X") ~ "Protein",
      Parameter %in% c("H_Y","Sigma_Y") ~ "mRNA",
      TRUE ~ "Slope"
    ),
    Parameter.type = case_when(
      Parameter %in% c("H_X","H_Y") ~ "Selection",
      Parameter %in% c("Sigma_X","Sigma_Y") ~ "Drift",
      TRUE ~ "Slope"
    ))

param.table.w.ss <- param.table.w.ss %>%
  right_join(bin.ss.div,by=c("Data" = "Bin"))

```

```{r}
ss.div.per.nt.per.bin <- param.table.w.ss %>% 
  filter(Model_Fit == "Protein_selection_strong" & Parameter.type == "Selection") %>%
  dplyr::select(Data,Mean,Abundance,Distance,Median.SS.Diverge) %>%
  nest(Data,Mean,Median.SS.Diverge) %>%
  mutate(cor = purrr::map(data,~cor(.x$Mean,.x$Median.SS.Diverge,use="pairwise.complete.obs",method="spearman"))) %>%
  unnest(cor)


sel.vs.mrna.ss.diverge.all.nt <- ggplot(ss.div.per.nt.per.bin,aes(x=Distance,y=cor)) +
  geom_point(aes(color=Abundance)) +
  geom_smooth(aes(color=Abundance)) +
  scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  ylab("Spearman correlation:\nRate of adaptation vs. free energy divergence") +
  xlab("Distance from start codon (nt)") +
  geom_vline(xintercept = 0,linetype="dashed")+
  labs(color = "Expression type") +
  theme_cowplot()

sel.vs.mrna.ss.diverge.all.nt
```

#### Based on Chew et al. transcripts

Repeating the analysis of mRNA secondary structures, but using transcripts as identified in Chew et al Nat. Comm. 2016.   

```{r}
hsapien.prot <- read_tsv("Data/PaxDB/paxdb_human_whole_organism_prot_abundance_integrated.tsv",comment="#") %>%
  mutate(string_external_id = str_remove(string_external_id,"9606."))

ensembl.prot.to.gene <- read_tsv("Data/PaxDB/hsapien_gene_mapping.tsv")

hsapien.prot <- hsapien.prot %>%
  left_join(ensembl.prot.to.gene,by=c("string_external_id"="ensembl_peptide_id"))

ba.data <- read_tsv("Data/mean_se_rna_protein.tsv")
ba.genes <- ba.data %>%
  dplyr::select(Gene_ID) %>%
  deframe() %>%
  unique()

hsapien.prot.filt <- hsapien.prot %>%
  filter(ensembl_gene_id %in% ba.genes) %>% 
  group_by(ensembl_gene_id) %>% 
  summarize(abundance = sum(abundance))

missing.ids <- ba.genes[which(!ba.genes %in% hsapien.prot.filt$ensembl_gene_id)]

hsapien.prot.filt <- hsapien.prot.filt %>%
  mutate(log10.abundance = log10(abundance),
         Bin = Hmisc::cut2(log10.abundance,g = 10))


start.codon.mrna.struct <- read_tsv("Data/Translation_efficiency/mrna_secondary_struct/human_vs_mouse_start_codon_ss_chew_etal_transcripts.tsv") %>%
  mutate(SS.Divergence = abs(CDS_SS_EFE_Human - CDS_SS_EFE_Mouse)) %>%
  filter(!is.na(SS.Divergence) & !is.na(ensembl_gene_id))

hsapien.prot.filt.ss.div <- hsapien.prot.filt %>%
  dplyr::right_join(start.codon.mrna.struct,by="ensembl_gene_id") %>%
  filter(!is.na(Bin))

bin.ss.div <- hsapien.prot.filt.ss.div %>%
  group_by(Bin,Distance) %>%
  summarize(Median.SS.Diverge = median(SS.Divergence,na.rm=T))

new.labels <- c("0-10","10-20","20-30","30-40","40-50","50-60","60-70","70-80","80-90","90-100")
names(new.labels) <- levels(bin.ss.div$Bin)
bin.ss.div$Bin <- unname(new.labels[bin.ss.div$Bin])

param.table.w.ss <- prot.dom.param.table %>%
    filter((Parameter %in% c("H_Y","H_X","Q","Sigma_Y","Sigma_X"))  & Model_Fit == "Protein_selection_strong") %>%
    mutate(Abundance = case_when(
      Parameter %in% c("H_X","Sigma_X") ~ "Protein",
      Parameter %in% c("H_Y","Sigma_Y") ~ "mRNA",
      TRUE ~ "Slope"
    ),
    Parameter.type = case_when(
      Parameter %in% c("H_X","H_Y") ~ "Selection",
      Parameter %in% c("Sigma_X","Sigma_Y") ~ "Drift",
      TRUE ~ "Slope"
    ))

param.table.w.ss <- param.table.w.ss %>%
  right_join(bin.ss.div,by=c("Data" = "Bin"))

```
```{r}
ss.div.per.nt.per.bin <- param.table.w.ss %>% 
  filter(Model_Fit == "Protein_selection_strong" & Parameter.type == "Selection") %>%
  dplyr::select(Data,Mean,Abundance,Distance,Median.SS.Diverge) %>%
  nest(Data,Mean,Median.SS.Diverge) %>%
  mutate(cor = purrr::map(data,~cor(.x$Mean,.x$Median.SS.Diverge,use="pairwise.complete.obs",method="spearman"))) %>%
  unnest(cor)


sel.vs.mrna.ss.diverge.all.nt.chew <- ggplot(ss.div.per.nt.per.bin,aes(x=Distance,y=cor)) +
  geom_point(aes(color=Abundance)) +
  geom_smooth(aes(color=Abundance)) +
  scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  ylab("Spearman correlation:\nRate of adaptation vs. free energy divergence") +
  xlab("Distance from start codon (nt)") +
  geom_vline(xintercept = 0,linetype="dashed")+
  labs(color = "Expression type") +
  theme_cowplot()

sel.vs.mrna.ss.diverge.all.nt.chew

ggsave2("Figures/mrna_ss_divergence_using_chew_etal_transcripts.pdf",plot = sel.vs.mrna.ss.diverge.all.nt.chew,width=7,height=5)

```


### 5'-UTR PhyloP scores

```{r}
sel.vs.phylo.p <- ggplot(param.table.w.enh.eqtl %>% filter(Parameter.type=="Selection") ,aes(x=Median.PhyloP,y=Mean,color=Abundance)) +
  geom_point() +
  xlab("Human 5'-UTR PhyloP scores") +
  ylab(expression("Rate of adaptation (Log scale)")) +
  geom_errorbar(aes(ymin=LB,ymax=UB),width=0) +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  theme_cowplot() + 
  labs(color="Expression type") +
  facet_wrap(~Abundance)

sel.vs.phylo.p


sel.vs.phast.p <- ggplot(param.table.w.enh.eqtl %>% filter(Parameter.type=="Selection") ,aes(x=Median.Phast,y=Mean,color=Abundance)) +
  geom_point() +
  xlab("Human 5'-UTR phastCons scores") +
  ylab(expression("Rate of adaptation (Log scale)")) +
  geom_errorbar(aes(ymin=LB,ymax=UB),width=0) +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  scale_color_manual(values=c("mRNA"=mrna.color,"Protein"=prot.color)) +
  theme_cowplot() + 
  labs(color="Expression type") +
  facet_wrap(~Abundance)

sel.vs.phast.p

seq.scores <- (sel.vs.phast.p / sel.vs.phylo.p) + plot_annotation(tag_levels = "A") + plot_layout(guides = "collect")

ggsave2("Figures/adaptation_vs_phylop.pdf",seq.scores,height = 8,width=8)


```


### 5'-UTR and translation efficiency

```{r}

a.r.slope <- prot.dom.param.table %>% 
  filter(Model_Fit=="Protein_selection_strong" & Parameter == "Q")

te.opt.per.bin <- prot.dom.param.table %>% 
  filter(Model_Fit=="Protein_selection_strong" & str_detect(Parameter,"Psi")) %>% 
  separate(Parameter,sep = "(?<=[XY])_",into=c("Parameter","Gene")) %>%
  pivot_wider(id_cols=c(Data,Gene),names_from=Parameter,values_from=Mean) %>% 
  left_join(a.r.slope,by="Data") %>%
  mutate(Protein_opt = Psi_X, 
         mRNA_opt = Psi_Y + Mean * Psi_X,
         TE_opt = Protein_opt - mRNA_opt,
         Theta.Ratio = Psi_X - Psi_Y) %>%
  group_by(Data) %>%
  summarize(Median.Theta.P = median(Psi_X),
            Median.Theta.R = median(Psi_Y),
            Median.Opt.TE = median(TE_opt),
            Median.Theta.Ratio = median(Theta.Ratio),
            a_R = mean(Mean),
            a_R.Theta.P = median(a_R * Psi_X))

te.opt.per.bin <- te.opt.per.bin %>% 
  left_join(human.bin.features,by=c("Data" = "Bin"))


opt.te.vs.human.te <- ggplot(te.opt.per.bin,aes(x=Median.Opt.TE,y=Median.Log.TE)) +
  geom_point() +
  theme_cowplot() +
  xlab("Median Log(Optimal TE)") +
  ylab("Median Human Log(TE)") +
  geom_abline(intercept=0,slope=1,linetype="dashed") +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) 
opt.te.vs.human.te

ggsave2("Figures/ribo_te_vs_optimum_te.pdf",opt.te.vs.human.te,width=4,height = 4)

```

```{r}
binned.expr <- list.files("Data/Bin_based_on_paxdb_human_prot_abund/10_groups/",full.names = T)
df <- purrr::map(binned.expr,read_tsv,col_types=cols())
names(df) <- c("0-10","10-20","20-30","30-40","40-50","50-60","60-70","70-80","80-90","90-100")
test <- lapply(df,function(x){
  x <- x %>%
    split(f=as.factor(.$Species))
  slopes <- lapply(x,function(y)
    {
    lmodel2::lmodel2(Mean_RNA ~ Mean_Protein,data=y,range.x="interval",range.y="interval") %>% broom::tidy()
  }) %>% bind_rows(.id="Species")
}) %>% bind_rows(.id="Data")

test.rma <- test %>%
  filter(method=="RMA" & term == "Slope" & Species != "Homo_sapiens")

summary.rma <- test.rma %>%
  group_by(Data) %>%
  summarize(Mean.Slope = mean(estimate),
            SD.Slope = sd(estimate))

summary.rma <- summary.rma %>%
  left_join(param.table.q)

rma.vs.ar <- ggplot(summary.rma,aes(y=Mean,x=Mean.Slope)) +
  geom_point(color=slope.color) +
  #ggrepel::geom_text_repel() +
  geom_errorbar(aes(ymin=LB,ymax=UB),height=0,color=slope.color) +
  geom_errorbarh(aes(xmin=Mean.Slope - SD.Slope,xmax=Mean.Slope + SD.Slope),height=0,color=slope.color) +
  theme_cowplot() +
  ylab(expression("a"[R])) +
  xlab("Mean Ranged Major Axis Slope\nR ~ P") +
  stat_cor(method="spearman",aes(label=..r.label..)) 
rma.vs.ar


```

## 5'-UTR length 

```{r}

slope.vs.utr.human <- ggplot(param.table.w.enh.eqtl %>% filter(Parameter.type=="Slope") ,aes(x=Median.UTR5.Human,y=Mean)) +
  geom_point(color=slope.color) +
  xlab("Median length of 5'-UTRs") +
  ylab(expression("Evolutionary slope a"[R])) +
  geom_errorbar(aes(ymin=LB,ymax=UB),width=0,color=slope.color) +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  theme_cowplot()


slope.vs.utr.mouse <- ggplot(param.table.w.enh.eqtl %>% filter(Parameter.type=="Slope") ,aes(x=Median.UTR5.Mouse,y=Mean)) +
  geom_point(color=slope.color) +
  xlab("Median length of 5'-UTRs") +
  ylab(expression("Evolutionary slope a"[R])) +
  geom_errorbar(aes(ymin=LB,ymax=UB),width=0,color=slope.color) +
  stat_cor(method="spearman",aes(label=..r.label..),show.legend = F) +
  theme_cowplot()



human.mouse.utr.length <- slope.vs.utr.human | slope.vs.utr.mouse + plot_annotation(tag_levels = "A")

ggsave2("Figures/human_mouse_utr_length.pdf",human.mouse.utr.length,width=10,height = 5)
```



## Re-ordering figures for Science submission - will not work unless you were able to generate the histograms for the parameters. 

```{r fig.width=9,fig.height=9}
layout <- "
AABB
CCDD
"

fig.3 <- dic.bar + alpha.plot + sel.vs.gene.exp.p + ess.p + plot_annotation(tag_levels="A") + plot_layout(design = layout)

ggsave2("Figures/Figure_3_science.pdf",fig.3,width=12,height = 9)

```

```{r fig.width=9,fig.height=10}
layout <- "
AABB
CCDD
EEFF
"

fig.4 <- sel.vs.enh + sel.vs.mrna.ss.diverge.all.nt + slope.plot + q.p + rma.vs.ar + plot_annotation(tag_levels="A") + plot_layout(design = layout)
fig.4
ggsave2("Figures/Figure_4_science.pdf",fig.4,width=12,height = 12.5)

```







